# 🏦 Customer Churn & Retention Analytics


**Objective:**  
Analyze customer data to identify the key factors associated with customer churn, understand customer segments, and provide actionable retention recommendations.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("/content/Churn_Modelling.csv")

In [3]:
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [4]:
df.shape

(10000, 14)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [6]:
df.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


In [8]:
df.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

**Data quality check**

In [9]:
df.isnull().sum()

,0
RowNumber,0
CustomerId,0
Surname,0
CreditScore,0
Geography,0
Gender,0
Age,0
Tenure,0
Balance,0
NumOfProducts,0


In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.nunique()

,0
RowNumber,10000
CustomerId,10000
Surname,2932
CreditScore,460
Geography,3
Gender,2
Age,70
Tenure,11
Balance,6382
NumOfProducts,4


In [12]:
df["Exited"].value_counts()

,count
Exited,
0,7963
1,2037


In [13]:
df["Exited"].value_counts(normalize=True)*100

,proportion
Exited,
0,79.63
1,20.37


**Data Quality Analysis**

In [14]:
print("Missing Values:")
print(df.isnull().sum())

print("\n" + "="*50)

print("Duplicate Rows:")
print(df.duplicated().sum())

print("\n" + "="*50)

print("Data Types:")
print(df.dtypes)

Missing Values:
RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

Duplicate Rows:
0

Data Types:
RowNumber            int64
CustomerId           int64
Surname             object
CreditScore          int64
Geography           object
Gender              object
Age                  int64
Tenure               int64
Balance            float64
NumOfProducts        int64
HasCrCard            int64
IsActiveMember       int64
EstimatedSalary    float64
Exited               int64
dtype: object


In [15]:
categorical_columns = [
    "Geography",
    "Gender",
    "NumOfProducts",
    "HasCrCard",
    "IsActiveMember",
    "Exited"
]

for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts())


--- Geography ---
Geography
France     5014
Germany    2509
Spain      2477
Name: count, dtype: int64

--- Gender ---
Gender
Male      5457
Female    4543
Name: count, dtype: int64

--- NumOfProducts ---
NumOfProducts
1    5084
2    4590
3     266
4      60
Name: count, dtype: int64

--- HasCrCard ---
HasCrCard
1    7055
0    2945
Name: count, dtype: int64

--- IsActiveMember ---
IsActiveMember
1    5151
0    4849
Name: count, dtype: int64

--- Exited ---
Exited
0    7963
1    2037
Name: count, dtype: int64


In [16]:
geography_churn = (
    df.groupby("Geography")["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

geography_churn.columns = [
    "Geography",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

geography_churn["Churn_Rate"] *= 100

geography_churn.sort_values(
    "Churn_Rate",
    ascending=False
)

,Geography,Total_Customers,Churned_Customers,Churn_Rate
1,Germany,2509,814,32.443204
2,Spain,2477,413,16.673395
0,France,5014,810,16.154767


### Key Finding — Geography

Germany has the highest customer churn rate at 32.44%, which is approximately
twice the churn rate observed in France (16.15%) and Spain (16.67%).

In [17]:
df["Age_Group"] = pd.cut(
    df["Age"],
    bins=[17, 25, 35, 45, 55, 65, 100],
    labels=[
        "18-25",
        "26-35",
        "36-45",
        "46-55",
        "56-65",
        "66+"
    ]
)

age_churn = (
    df.groupby("Age_Group", observed=False)["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

age_churn.columns = [
    "Age_Group",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

age_churn["Churn_Rate"] *= 100

age_churn

,Age_Group,Total_Customers,Churned_Customers,Churn_Rate
0,18-25,611,46,7.528642
1,26-35,3542,301,8.498024
2,36-45,3736,733,19.619914
3,46-55,1311,663,50.572082
4,56-65,536,259,48.320896
5,66+,264,35,13.257576


##Key Finding — Age
Customer churn varies substantially across age groups.

Customers aged 46-55 have the highest churn rate at 50.57%, followed by customers aged 56-65 at 48.32%. In comparison, younger customers aged
18-25 and 26-35 have much lower churn rates of 7.53% and 8.50%.

In [18]:
gender_churn = (
    df.groupby("Gender")["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

gender_churn.columns = [
    "Gender",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

gender_churn["Churn_Rate"] *= 100

gender_churn.sort_values("Churn_Rate", ascending=False)

,Gender,Total_Customers,Churned_Customers,Churn_Rate
0,Female,4543,1139,25.071539
1,Male,5457,898,16.455928


## Key Finding — Gender

Female customers have a higher observed churn rate (25.07%) than male
customers (16.46%), representing an 8.62 percentage-point difference.

In [19]:
activity_churn = (
    df.groupby("IsActiveMember")["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

activity_churn.columns = [
    "IsActiveMember",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

activity_churn["Churn_Rate"] *= 100

activity_churn

,IsActiveMember,Total_Customers,Churned_Customers,Churn_Rate
0,0,4849,1302,26.850897
1,1,5151,735,14.269074


## Key Finding — Customer Activity

Inactive customers have a substantially higher observed churn rate (26.85%)
than active customers (14.27%), representing a 12.58 percentage-point
difference.

In [20]:
credit_card_churn = (
    df.groupby("HasCrCard")["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

credit_card_churn.columns = [
    "HasCrCard",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

credit_card_churn["Churn_Rate"] *= 100

credit_card_churn

,HasCrCard,Total_Customers,Churned_Customers,Churn_Rate
0,0,2945,613,20.814941
1,1,7055,1424,20.184266


## Key Finding — Credit Card Ownership

Churn rates are very similar between customers with and without a credit
card. Customers without a credit card have a churn rate of 20.81%, compared
with 20.18% among customers with a credit card.

In [21]:
tenure_churn = (
    df.groupby("Tenure")["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

tenure_churn.columns = [
    "Tenure",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

tenure_churn["Churn_Rate"] *= 100

tenure_churn

,Tenure,Total_Customers,Churned_Customers,Churn_Rate
0,0,413,95,23.002421
1,1,1035,232,22.415459
2,2,1048,201,19.179389
3,3,1009,213,21.110010
4,4,989,203,20.525784
5,5,1012,209,20.652174
6,6,967,196,20.268873
7,7,1028,177,17.217899
8,8,1025,197,19.219512
9,9,984,213,21.646341


## Key Finding — Customer Tenure

Churn rates remain relatively stable across different customer tenure
levels, ranging from 17.22% to 23.00%

In [22]:
df["CreditScore_Group"] = pd.cut(
    df["CreditScore"],
    bins=[300, 580, 670, 740, 851],
    labels=["Poor", "Fair", "Good", "Excellent"]
)

credit_churn = (
    df.groupby("CreditScore_Group", observed=False)["Exited"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

credit_churn.columns = [
    "CreditScore_Group",
    "Total_Customers",
    "Churned_Customers",
    "Churn_Rate"
]

credit_churn["Churn_Rate"] *= 100

credit_churn

,CreditScore_Group,Total_Customers,Churned_Customers,Churn_Rate
0,Poor,2393,530,22.147931
1,Fair,3350,683,20.388060
2,Good,2397,451,18.815186
3,Excellent,1860,373,20.053763


## Key Finding — Credit Score

Churn rates are relatively similar across credit score groups, ranging from
18.82% for customers with Good credit scores to 22.15% for customers with
Poor credit scores.

## Key EDA Insights

The exploratory analysis identified several important patterns associated
with customer churn:

1. **Age is a major churn-related factor.**
   Customers aged 46-55 have the highest churn rate at 50.57%, followed by
   the 56-65 group at 48.32%.

2. **Customer activity is strongly associated with retention.**
   Inactive customers have a churn rate of 26.85%, compared with 14.27% for
   active customers.

3. **Geography shows a significant difference in churn.**
   Germany has the highest churn rate at 32.44%, approximately twice the
   churn rate observed in France and Spain.

4. **Gender shows a noticeable difference.**
   Female customers have a churn rate of 25.07%, compared with 16.46% for
   male customers.

5. **Credit score appears to have a relatively weak relationship with churn.**
   Churn rates across credit-score groups remain relatively close.

6. **Customer tenure does not show a clear trend.**
   Churn rates fluctuate between 17.22% and 23.00% across tenure levels.

7. **Credit-card ownership has minimal association with churn.**
   Churn rates differ by only 0.63 percentage points between customers with
   and without a credit card.

These findings suggest that retention efforts should primarily investigate
customer age, engagement/activity, and geography, while treating other
variables as supporting factors rather than standalone churn indicators.
